File: leetcode-solutions/ruby/140. Word Break II.rb
# https://leetcode.com/problems/word-break-ii/

# @param {String} s
# @param {String[]} word_dict
# @return {String[]}
def word_break(s, word_dict)
    dict = {}
    word_dict.each do |word|
        dict[word] = true
    end
    memo = {}
    get_valid_words(dict, s, memo)
end

def get_valid_words(dict, string, memo)
    return [] if string.empty?
    if memo.key? string
        return memo[string]
    end
    
    size = string.size
    words = []
    size.times do |i|
        sub_string = string[0..i]
        
        if dict.key?(sub_string)
            if i == size - 1
                words << sub_string
                next
            end
            
            strings = get_valid_words(dict, string[i+1..-1], memo)
            unless strings.empty?
                words += strings.map{|s| sub_string + " " + s}
            end
        end
    end
    memo[string] = words
    memo[string]
end

File: leetcode-solutions/python/140. Word Break II.py
# https://leetcode.com/problems/word-break-ii/

from collections import defaultdict


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        memo = defaultdict()
        memo[''] = []
        return self.get_sentences(s, wordDict, memo)

    def get_sentences(self, text, wordDict, memo):
        if text in memo:
            return memo[text]

        sentences = []
        for idx in range(len(text)):
            sub_str = text[0:idx+1]

            if sub_str in wordDict:
                # handle last part of the sentence
                if idx == len(text) - 1:
                    sentences.append(sub_str)
                    continue

                partial_sentences = self.get_sentences(
                    text[idx+1:], wordDict, memo)
                temp_sentences = [sub_str + " " + s for s in partial_sentences]
                sentences += temp_sentences

        memo[text] = sentences
        return sentences
